### PR TITLE
fixed ipv6 address join ipv4 group failed

### LINKED ISSUE #10995: Netty UDP multicast using Epoll throws Exception while joining multicast group
### Expected behavior
Joining multicast group on given port
### Actual behavior
Throws Exception:
```
Exception in thread "main" io.netty.channel.ChannelException: setsockopt() failed: Invalid argument
	at io.netty.channel.epoll.LinuxSocket.joinGroup(Native Method)
	at io.netty.channel.epoll.LinuxSocket.joinGroup(LinuxSocket.java:124)
	at io.netty.channel.epoll.EpollDatagramChannel.joinGroup(EpollDatagramChannel.java:172)
	at io.netty.channel.epoll.EpollDatagramChannel.joinGroup(EpollDatagramChannel.java:154)
	at io.netty.channel.epoll.EpollDatagramChannel.joinGroup(EpollDatagramChannel.java:147)
	at trading.app.udp.UDPEpollListener.run(UDPEpollListener.java:56)
	at trading.app.udp.UDPEpollListener.main(UDPEpollListener.java:33)
```
After changing all Epoll classes to NIO the code runs without any problem.
### Steps to reproduce
Run the given code
### Minimal yet complete reproducer code (or URL to code)
```
import io.netty.bootstrap.Bootstrap;
import io.netty.channel.ChannelFuture;
import io.netty.channel.ChannelFutureListener;
import io.netty.channel.ChannelInitializer;
import io.netty.channel.ChannelOption;
import io.netty.channel.epoll.EpollDatagramChannel;
import io.netty.channel.epoll.EpollEventLoopGroup;
import lombok.val;

import java.net.Inet6Address;
import java.net.InetSocketAddress;
import java.net.NetworkInterface;
import java.net.SocketException;

public class UDPEpollListener {
    private final NetworkInterface networkInterface;
    private final Inet6Address networkInterfaceAddress;
    private int socketBufferSize;
    private String multicastGroupIpv4Address;
    private int multicastGroupPort;

    public UDPEpollListener(String networkInterfaceName, int socketBufferSize, String multicastGroupIpv4Address, int multicastGroupPort) {
        this.networkInterface = getNetworkInterface(networkInterfaceName);
        this.networkInterfaceAddress = getLocalAddress(this.networkInterface);
        this.socketBufferSize = socketBufferSize;
        this.multicastGroupIpv4Address = multicastGroupIpv4Address;
        this.multicastGroupPort = multicastGroupPort;
    }

    public static void main(String[] args) {
        val listener = new UDPEpollListener("enp0s3", 16777216, "233.96.6.2", 23455);
        try {
            listener.run();
        } catch (InterruptedException e) {
            e.printStackTrace();
        }
    }

    private void run() throws InterruptedException {
        Bootstrap bootstrap = new Bootstrap();
        bootstrap
                .group(new EpollEventLoopGroup())
                .channel(EpollDatagramChannel.class)
                .option(ChannelOption.SO_REUSEADDR, true)
                .option(ChannelOption.IP_MULTICAST_IF, this.networkInterface)
                .option(ChannelOption.SO_RCVBUF, this.socketBufferSize)
                .localAddress(this.networkInterfaceAddress, multicastGroupPort)
                .handler(new ChannelInitializer<EpollDatagramChannel>() {
                    @Override
                    protected void initChannel(EpollDatagramChannel datagramChannel) {
                        datagramChannel.pipeline().addLast(new MulticastHandler());
                    }
                });
        EpollDatagramChannel channel = (EpollDatagramChannel) bootstrap.bind(multicastGroupPort).sync().channel();
        InetSocketAddress multicastSocketAddress = new InetSocketAddress(multicastGroupIpv4Address, multicastGroupPort);
        ChannelFuture channelFuture = channel.joinGroup(multicastSocketAddress, this.networkInterface).sync();
        channelFuture.addListener((ChannelFutureListener) listener -> {
            if (channelFuture.isSuccess()) {
                System.out.println("Working...");
            }});
    }

    private NetworkInterface getNetworkInterface(String networkInterfaceName) {
        try {
            return NetworkInterface.getByName(networkInterfaceName);
        } catch (SocketException e) {
            throw new RuntimeException();
        }
    }

    private Inet6Address getLocalAddress(NetworkInterface networkInterface) {
        val addresses = networkInterface.getInetAddresses();
        while (addresses.hasMoreElements()) {
            val address = addresses.nextElement();
            if (address instanceof Inet6Address) {
                return (Inet6Address) address;
            }
        }
        throw new RuntimeException();
    }
}
```
### Netty version
4.1.58.Final
### JVM version (e.g. `java -version`)
openjdk version "15.0.2" 2021-01-19
OpenJDK Runtime Environment (build 15.0.2+7-27)
OpenJDK 64-Bit Server VM (build 15.0.2+7-27, mixed mode, sharing)
### OS version (e.g. `uname -a`)
Linux centos.dev 4.18.0-240.10.1.el8_3.x86_64 #1 SMP Mon Jan 18 17:05:51 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux

### PR DESCRIPTION
**Pull request by wineway**
Motivation:

#10995 

when `io.netty.channel.unix.Socket` is ipv6 and join a multicast group with ipv4 address will cause `io.netty.channel.ChannelException: setsockopt() failed: Invalid argument` (at least in `Linux centos.dev 4.18.0-240.10.1.el8_3.x86_64 #1 SMP Mon Jan 18 17:05:51 UTC 2021 x86_64 x86_64 x86_64 GNU/Linux`)

Modification:

check if target group address is ipv6 before call  `io.netty.channel.epoll.LinuxSocket#joinGroup(java.net.InetAddress, java.net.NetworkInterface, java.net.InetAddress)` 

I'm not sure if this modification is currect, but i checked source code of java NIO

```
Java_sun_nio_ch_Net_canJoin6WithIPv4Group0(JNIEnv* env, jclass cl)
{
#if defined(__APPLE__)
    /* IPV6_ADD_MEMBERSHIP can be used to join IPv4 multicast groups */
    return JNI_TRUE;
#else
    /* IPV6_ADD_MEMBERSHIP cannot be used to join IPv4 multicast groups */
    return JNI_FALSE;
#endif
}
```
seems ipv6 address can't join ipv4 group except osx

Result:

test on `Linux 3.10.0-514.el7.x86_64 #1 SMP Tue Nov 22 16:42:41 UTC 2016 x86_64 x86_64 x86_64 GNU/Linux` exception  ` setsockopt() failed: Invalid argument` has fixed

Fixes #10995

### CODE REVIEW

**Comment by normanmaurer** [comment]
@wineway did you verify this really fixed the issue ? Maybe you could also add a unit test ?

**Comment by wineway** [comment]
> @wineway did you verify this really fixed the issue ? Maybe you could also add a unit test ?

of cource my manual test has passed, and i would try to add proper unit test

I can explain why this modification can work;

refer to java nio implement; in linux, ipv6 socket can join ipv4 multicast group
```
Java_sun_nio_ch_Net_canIPv6SocketJoinIPv4Group0(JNIEnv* env, jclass cl)
{
#if defined(__linux__) || defined(__APPLE__)
    /* IPv6 sockets can join IPv4 multicast groups */
    return JNI_TRUE;
#else
    /* IPv6 sockets cannot join IPv4 multicast groups */
    return JNI_FALSE;
#endif
}
```
but  `IPV6_ADD_MEMBERSHIP` can't be used to join ipv4
```
Java_sun_nio_ch_Net_canJoin6WithIPv4Group0(JNIEnv* env, jclass cl)
{
#if defined(__APPLE__)
    /* IPV6_ADD_MEMBERSHIP can be used to join IPv4 multicast groups */
    return JNI_TRUE;
#else
    /* IPV6_ADD_MEMBERSHIP cannot be used to join IPv4 multicast groups */
    return JNI_FALSE;
#endif
}
```
so we should use `IP_ADD_MEMBERSHIP` instead of `IPV6_ADD_MEMBERSHIP` or `IPV6_JOIN_GROUP` 

the original  `bool ipv6` depend on if local socket is instance of ipv6,
when ipv4 group pass in, it will initialize as `sockaddr_in6` by `netty_unix_socket_initSockaddr` function, and will be set option `IPV6_JOIN_GROUP`, this will cause an exception

and in fact, when `isIpv6 == true` , no matter `group addr` or `interface addr` is `Inet4Adder`, we shouldn't pass  `ipv6 = true` to `netty_epoll_linuxsocket_joinGroup`


and alse, in `joinSsmGroup` , java nio implement will check if  `source` is same type with `group`

```
        if (source != null) {
            if (source.isAnyLocalAddress())
                throw new IllegalArgumentException("Source address is a wildcard address");
            if (source.isMulticastAddress())
                throw new IllegalArgumentException("Source address is multicast address");
            if (source.getClass() != group.getClass())
                throw new IllegalArgumentException("Source address is different type to group");
        }
```

but netty implement didn't do this, I am not sure if we should add this check or just let it go

**Comment by normanmaurer** [comment]
@wineway did you sign our icla ?

https://netty.io/s/icla

**Comment by wineway** [comment]
> @wineway did you sign our icla ?
> 
> https://netty.io/s/icla

@normanmaurer sure, I have add tests, PTAL

thx

**Comment by normanmaurer** [comment]
@wineway the unit test fails... please check

File: transport-native-epoll/src/test/java/io/netty/channel/epoll/EpollDatagramMulticastIpv6WithIpv4AddrTest.java [hunk 1]
```diff
@@ -0,0 +1,36 @@
+/*
+ * Copyright 2019 The Netty Project
```
Comment (normanmaurer) [inline transport-native-epoll/src/test/java/io/netty/channel/epoll/EpollDatagramMulticastIpv6WithIpv4AddrTest.java:2]: 2021

Comment (wineway) [reply transport-native-epoll/src/test/java/io/netty/channel/epoll/EpollDatagramMulticastIpv6WithIpv4AddrTest.java]: fixed

**Review by normanmaurer — APPROVED** [review]

**Review by chrisvest — APPROVED** [review]

**Comment by chrisvest** [comment]
@wineway Thanks!

**Pull request merged**